logical/plugin/grpc_backend_client.go

```diff
@@ -0,0 +1,246 @@
+	server *grpc.Server
+	// clientConn is the underlying grpc connection to the server, we store it
+	// so it can be cleaned up.
+	clientConn *grpc.ClientConn
+	doneCtx    context.Context
+}
+
+func (b *backendGRPCPluginClient) HandleRequest(ctx context.Context, req *logical.Request) (*logical.Response, error) {
+	if b.metadataMode {
+		return nil, ErrClientInMetadataMode
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	quitCh := pluginutil.CtxCancelIfCanceled(cancel, b.doneCtx)
+	defer close(quitCh)
+	defer cancel()
+
+	protoReq, err := pb.LogicalRequestToProtoRequest(req)
+	if err != nil {
+		return nil, err
+	}
+
+	reply, err := b.client.HandleRequest(ctx, &pb.HandleRequestArgs{
+		Request: protoReq,
+	})
+	if err != nil {
+		if b.doneCtx.Err() != nil {
+			return nil, ErrPluginShutdown
+		}
+
+		return nil, err
+	}
+	resp, err := pb.ProtoResponseToLogicalResponse(reply.Response)
+	if err != nil {
+		return nil, err
+	}
+	if reply.Err != nil {
+		return resp, pb.ProtoErrToErr(reply.Err)
+	}
+
+	return resp, nil
+}
+
+func (b *backendGRPCPluginClient) SpecialPaths() *logical.Paths {
+	// Timeout the connection
+	reply, err := b.client.SpecialPaths(b.doneCtx, &pb.Empty{})
+	if err != nil {
+		return nil
+	}
+
+	return &logical.Paths{
+		Root:            reply.Paths.Root,
+		Unauthenticated: reply.Paths.Unauthenticated,
+		LocalStorage:    reply.Paths.LocalStorage,
+		SealWrapStorage: reply.Paths.SealWrapStorage,
+	}
+}
+
+// System returns vault's system view. The backend client stores the view during
+// Setup, so there is no need to shim the system just to get it back.
+func (b *backendGRPCPluginClient) System() logical.SystemView {
+	return b.system
+}
+
+// Logger returns vault's logger. The backend client stores the logger during
+// Setup, so there is no need to shim the logger just to get it back.
+func (b *backendGRPCPluginClient) Logger() log.Logger {
+	return b.logger
+}
+
+func (b *backendGRPCPluginClient) HandleExistenceCheck(ctx context.Context, req *logical.Request) (bool, bool, error) {
+	if b.metadataMode {
+		return false, false, ErrClientInMetadataMode
+	}
+
+	protoReq, err := pb.LogicalRequestToProtoRequest(req)
+	if err != nil {
+		return false, false, err
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	quitCh := pluginutil.CtxCancelIfCanceled(cancel, b.doneCtx)
+	defer close(quitCh)
+	defer cancel()
+	reply, err := b.client.HandleExistenceCheck(ctx, &pb.HandleExistenceCheckArgs{
+		Request: protoReq,
+	})
+	if err != nil {
+		if b.doneCtx.Err() != nil {
+			return false, false, ErrPluginShutdown
+		}
+		return false, false, err
+	}
+	if reply.Err != nil {
+		return false, false, pb.ProtoErrToErr(reply.Err)
+	}
+
+	return reply.CheckFound, reply.Exists, nil
+}
+
+func (b *backendGRPCPluginClient) Cleanup() {
+	// Timout the connection incase we have a bad connection. We can't block on
+	// shutdown.
+	ctx, cancel := context.WithTimeout(b.doneCtx, time.Second)
```

Comment:
One second feels short. I could imagine that a plugin might need to close a number of connections or finalize resources.

Reply:
this should actually be removed and replaced with doneCtx. Will fix